File: spikes/ocaml/lib/store.ml
(** The SQL edge. Everything that knows about a database is here.

    The split is the same one {!Wire} makes for JSON: {!Issue.t} and {!Project.t} name no column and
    no query. What [store.mli] adds on top is the reason this file has one at all — it exports the
    loads and the writes, and keeps the row types, the column lists and the SQL private. The base
    tables cannot be named from outside, so every read is guaranteed to carry its liveness
    predicate.

    The queries are SQL text. That is one definition of the schema in [schema.sql] and a second in
    the column lists below, which is the price of an ORM that does not generate its rows from the
    schema — the same two definitions the Rust spike has. What it buys is that everything SQLite can
    say is sayable: [STRICT], [CHECK], a partial unique index, a join, mixed [ORDER BY] directions,
    [RETURNING].

    Blocking throughout. SQLite has no waiting in it, so a promise would be ceremony with nothing
    behind it. *)

open Caqti_request.Infix

type conn = (module Caqti_blocking.CONNECTION)
type error = Query_failed of string | Missing_after_insert of string

let show_error = function
  | Query_failed message -> message
  | Missing_after_insert what -> what ^ " vanished between the insert and the read"

let broken result = Result.map_error (fun e -> Error.Broken (show_error e)) result

(* Caqti's error is an open polymorphic variant spread over four constructors.
   It is narrowed to one closed type here, at the boundary, rather than being
   let out into the rest of the tree as a row type a caller has to match with
   [#Caqti_error.t as e]. *)
let failed e = Query_failed (Caqti_error.show e)
let ( let* ) = Result.bind

let exec (conn : conn) request params =
  let (module Db : Caqti_blocking.CONNECTION) = conn in
  Result.map_error failed (Db.exec request params)

let find (conn : conn) request params =
  let (module Db : Caqti_blocking.CONNECTION) = conn in
  Result.map_error failed (Db.find request params)

let find_opt (conn : conn) request params =
  let (module Db : Caqti_blocking.CONNECTION) = conn in
  Result.map_error failed (Db.find_opt request params)

let collect (conn : conn) request params =
  let (module Db : Caqti_blocking.CONNECTION) = conn in
  Result.map_error failed (Db.collect_list request params)

(* --- the schema ---------------------------------------------------------- *)

(** [PRAGMA foreign_keys] is per connection rather than per database, so it is not in [schema.sql]
    and cannot be. *)
let foreign_keys = Caqti_type.(unit ->. unit) ~oneshot:true "PRAGMA foreign_keys = ON"

let connect uri : (conn, error) result =
  match Caqti_blocking.connect (Uri.of_string uri) with
  | Error e -> Error (failed e)
  | Ok conn -> Result.map (fun () -> conn) (exec conn foreign_keys ())

(** One statement per request is Caqti's rule, so the file is split on its statement terminator. A
    [;] inside a comment would break this, which is why there is not one. *)
let statements sql =
  String.split_on_char ';' sql |> List.map String.trim |> List.filter (fun s -> s <> "")

let initialise conn : (unit, error) result =
  List.fold_left
    (fun acc sql ->
      let* () = acc in
      exec conn (Caqti_type.(unit ->. unit) ~oneshot:true sql) ())
    (Ok ()) (statements Schema.ddl)

(* --- column types -------------------------------------------------------- *)

(** An enum's SQL type, from the same two matches its decoder and its schema are built from.
    [decode] is where a value the [CHECK] constraint would have prevented arrives; it cannot be
    anything but a read failure. *)
let enum_column ~name ~to_string ~of_string ~names =
  Caqti_type.enum ~encode:to_string
    ~decode:(fun wire ->
      match of_string wire with
      | Some value -> Ok value
      | None ->
          Error (Printf.sprintf "%s: %S is not one of %s" name wire (String.concat ", " names)))
    name

let project_status =
  enum_column ~name:"project_status" ~to_string:Project.Status.to_string
    ~of_string:Project.Status.of_string ~names:Project.Status.names

let issue_status =
  enum_column ~name:"issue_status" ~to_string:Issue.Status.to_string
    ~of_string:Issue.Status.of_string ~names:Issue.Status.names

(** [priority] is the one column whose SQL representation is not the wire name.
    {!Issue.Priority.to_int} is the same pair of values written as a match, which is what makes it
    exhaustive. *)
let priority =
  Caqti_type.custom Caqti_type.int
    ~encode:(fun value -> Ok (Issue.Priority.to_int value))
    ~decode:(fun n ->
      match Issue.Priority.of_int n with
      | Some value -> Ok value
      | None -> Error (Printf.sprintf "priority: %d is not 0 or 1" n))

(* --- rows ---------------------------------------------------------------- *)

(** The projects table as one row.

    Not {!Project.t}: it carries [deleted_at], which no live projection has, and not the counts,
    which are a second query. {!Caqti_type.product} builds this record directly, so each column
    names the field it fills — but the constructor's parameters are still positional, and swapping
    two same-typed columns between [columns] and the projections below compiles clean. That is the
    residue of the tuple decoder rather than its removal. *)
module Project_row = struct
  type t = {
    id : int;
    slug : string;
    title : string;
    body : string;
    status : Project.Status.t;
    created_at : string;
    updated_at : string;
    deleted_at : string option;
  }

  let columns = "id, slug, title, body, status, created_at, updated_at, deleted_at"

  let ty =
    Caqti_type.(
      product (fun id slug title body status created_at updated_at deleted_at ->
          { id; slug; title; body; status; created_at; updated_at; deleted_at })
      @@ proj int (fun r -> r.id)
      @@ proj string (fun r -> r.slug)
      @@ proj string (fun r -> r.title)
      @@ proj string (fun r -> r.body)
      @@ proj project_status (fun r -> r.status)
      @@ proj string (fun r -> r.created_at)
      @@ proj string (fun r -> r.updated_at)
      @@ proj (option string) (fun r -> r.deleted_at)
      @@ proj_end)
end

(** The issues table joined to its project, so [project_slug] arrives with the row rather than being
    paired up afterwards in OCaml. *)
module Issue_row = struct
  type t = {
    id : int;
    project_id : int;
    project_slug : string;
    title : string;
    body : string;
    status : Issue.Status.t;
    priority : Issue.Priority.t;
    status_note : string option;
    created_at : string;
    updated_at : string;
    deleted_at : string option;
  }

  let columns =
    "i.id, i.project_id, p.slug, i.title, i.body, i.status, i.priority, i.status_note, \
     i.created_at, i.updated_at, i.deleted_at"

  let ty =
    Caqti_type.(
      product
        (fun
          id
          project_id
          project_slug
          title
          body
          status
          priority
          status_note
          created_at
          updated_at
          deleted_at
        ->
          {
            id;
            project_id;
            project_slug;
            title;
            body;
            status;
            priority;
            status_note;
            created_at;
            updated_at;
            deleted_at;
          })
      @@ proj int (fun r -> r.id)
      @@ proj int (fun r -> r.project_id)
      @@ proj string (fun r -> r.project_slug)
      @@ proj string (fun r -> r.title)
      @@ proj string (fun r -> r.body)
      @@ proj issue_status (fun r -> r.status)
      @@ proj priority (fun r -> r.priority)
      @@ proj (option string) (fun r -> r.status_note)
      @@ proj string (fun r -> r.created_at)
      @@ proj string (fun r -> r.updated_at)
      @@ proj (option string) (fun r -> r.deleted_at)
      @@ proj_end)
end

(* --- reading ------------------------------------------------------------- *)

(** [created_at] is a second, so it is not a total order — two rows written in the same second tie,
    and SQLite is then free to return them either way round. The id breaks the tie. *)
let projects_query predicate =
  Printf.sprintf "SELECT %s FROM projects WHERE %s ORDER BY created_at, id" Project_row.columns
    predicate

let live_projects = Caqti_type.(unit ->* Project_row.ty) (projects_query "deleted_at IS NULL")

let live_project_by_slug =
  Caqti_type.(string ->? Project_row.ty) (projects_query "deleted_at IS NULL AND slug = ?")

let live_project_by_id =
  Caqti_type.(int ->? Project_row.ty) (projects_query "deleted_at IS NULL AND id = ?")

let deleted_projects =
  Caqti_type.(unit ->* Project_row.ty) (projects_query "deleted_at IS NOT NULL")

(** High priority first, oldest first within that. Two directions in one clause, in SQL, so nothing
    downstream has to remember to re-sort. *)
let issues_query predicate =
  Printf.sprintf
    "SELECT %s FROM issues i JOIN projects p ON p.id = i.project_id WHERE %s ORDER BY i.priority \
     DESC, i.created_at ASC, i.id ASC"
    Issue_row.columns predicate

let live_issues_of_project =
  Caqti_type.(string ->* Issue_row.ty)
    (issues_query "i.deleted_at IS NULL AND p.deleted_at IS NULL AND p.slug = ?")

let live_issue =
  Caqti_type.(int ->? Issue_row.ty)
    (issues_query "i.deleted_at IS NULL AND p.deleted_at IS NULL AND i.id = ?")

(** The one read that asks nothing of the project but that it be there. A row just inserted is
    loaded through this, and so is an issue in the trash whose project also went. *)
let issue_by_id = Caqti_type.(int ->? Issue_row.ty) (issues_query "i.id = ?")

let deleted_issues = Caqti_type.(unit ->* Issue_row.ty) (issues_query "i.deleted_at IS NOT NULL")

(** The counts, as one grouped query folded in OCaml.

    A [FILTER] clause per status would do it in SQL and put the three wire names in a fourth place.
    The fold is a match instead, which is exhaustive over the type. *)
let count_rows =
  Caqti_type.(unit ->* t3 int issue_status int)
    "SELECT project_id, status, COUNT(*) FROM issues WHERE deleted_at IS NULL GROUP BY project_id, \
     status"

let tally rows project_id =
  List.fold_left
    (fun (todo, doing, done_) (id, status, n) ->
      if id <> project_id then (todo, doing, done_)
      else
        match status with
        | Issue.Status.Todo -> (todo + n, doing, done_)
        | Issue.Status.Doing -> (todo, doing + n, done_)
        | Issue.Status.Done -> (todo, doing, done_ + n))
    (0, 0, 0) rows

let to_project counts (r : Project_row.t) : Project.t =
  let todo, doing, done_ = counts r.id in
  {
    id = r.id;
    slug = r.slug;
    title = r.title;
    body = r.body;
    status = r.status;
    todo;
    doing;
    done_;
    created_at = r.created_at;
    updated_at = r.updated_at;
  }

let to_issue (r : Issue_row.t) : Issue.t =
  {
    id = r.id;
    project_id = r.project_id;
    project_slug = r.project_slug;
    title = r.title;
    body = r.body;
    status = r.status;
    priority = r.priority;
    status_note = r.status_note;
    created_at = r.created_at;
    updated_at = r.updated_at;
  }

let counted conn =
  let* rows = collect conn count_rows () in
  Ok (tally rows)

let projects conn : (Project.t list, error) result =
  let* rows = collect conn live_projects () in
  let* counts = counted conn in
  Ok (List.map (to_project counts) rows)

let one_project conn request params =
  let* row = find_opt conn request params in
  match row with
  | None -> Ok None
  | Some row ->
      let* counts = counted conn in
      Ok (Some (to_project counts row))

let project ~slug conn : (Project.t option, error) result =
  one_project conn live_project_by_slug slug

let project_by_id id conn = one_project conn live_project_by_id id

let issues ~project_slug conn : (Issue.t list, error) result =
  let* rows = collect conn live_issues_of_project project_slug in
  Ok (List.map to_issue rows)

let issue ~id conn : (Issue.t option, error) result =
  let* row = find_opt conn live_issue id in
  Ok (Option.map to_issue row)

let issue_row conn id =
  let* row = find_opt conn issue_by_id id in
  Ok (Option.map to_issue row)

(** The trash is by the row's own [deleted_at] and nothing else.

    A project going does not put its issues here — it hides them, because {!issues} requires a live
    project. That is what makes restoring a project bring back exactly the issues that were not
    deleted in their own right: one row was written on the way out, so one row is cleared on the way
    back. *)
let trashed_projects conn : (Project.t Deleted.t list, error) result =
  let* rows = collect conn deleted_projects () in
  let* counts = counted conn in
  Ok
    (List.filter_map
       (fun (r : Project_row.t) ->
         Option.map
           (fun deleted_at -> { Deleted.inner = to_project counts r; deleted_at })
           r.deleted_at)
       rows)

let trashed_issues conn : (Issue.t Deleted.t list, error) result =
  let* rows = collect conn deleted_issues () in
  Ok
    (List.filter_map
       (fun (r : Issue_row.t) ->
         Option.map (fun deleted_at -> { Deleted.inner = to_issue r; deleted_at }) r.deleted_at)
       rows)

let emitted =
  Caqti_type.(unit ->* string) "SELECT sql FROM sqlite_master WHERE sql IS NOT NULL ORDER BY name"

let emitted_ddl conn : (string list, error) result = collect conn emitted ()

(* --- writing ------------------------------------------------------------- *)

let update_project_row =
  Caqti_type.(t5 string string project_status string int ->. unit)
    "UPDATE projects SET title = ?, body = ?, status = ?, updated_at = ? WHERE id = ?"

let update_issue_row =
  Caqti_type.(t7 string string issue_status priority (option string) string int ->. unit)
    "UPDATE issues SET title = ?, body = ?, status = ?, priority = ?, status_note = ?, updated_at \
     = ? WHERE id = ?"

let stamp_project_deleted =
  Caqti_type.(t3 string string int ->. unit)
    "UPDATE projects SET deleted_at = ?, updated_at = ? WHERE id = ?"

let stamp_issue_deleted =
  Caqti_type.(t3 string string int ->. unit)
    "UPDATE issues SET deleted_at = ?, updated_at = ? WHERE id = ?"

let clear_project_deleted =
  Caqti_type.(t2 string int ->. unit)
    "UPDATE projects SET deleted_at = NULL, updated_at = ? WHERE id = ?"

let clear_issue_deleted =
  Caqti_type.(t2 string int ->. unit)
    "UPDATE issues SET deleted_at = NULL, updated_at = ? WHERE id = ?"

let insert_project =
  Caqti_type.(t6 string string string project_status string string ->! int)
    "INSERT INTO projects (slug, title, body, status, created_at, updated_at) VALUES (?, ?, ?, ?, \
     ?, ?) RETURNING id"

let insert_issue =
  Caqti_type.(t7 int string string issue_status priority string string ->! int)
    "INSERT INTO issues (project_id, title, body, status, priority, created_at, updated_at) VALUES \
     (?, ?, ?, ?, ?, ?, ?) RETURNING id"

(** Persisting what an action returned. {!Action.run} produced the new value; this only writes it,
    so the enforcement point stays where {!Action} put it.

    [updated_at] is stamped here rather than by the action, which keeps every [execute] a pure
    function of the object and its payload — the thing that lets the whole action layer be tested
    with no clock and no database. *)
let update_project (project : Project.t) conn : (unit, error) result =
  exec conn update_project_row
    (project.title, project.body, project.status, Clock.now (), project.id)

let update_issue (issue : Issue.t) conn : (unit, error) result =
  exec conn update_issue_row
    ( issue.title,
      issue.body,
      issue.status,
      issue.priority,
      issue.status_note,
      Clock.now (),
      issue.id )

(** The soft delete, which is an update like any other — which is exactly why the type of what an
    action returned cannot say which of these to call, and why the group it was registered in has
    to. *)
let delete_project (project : Project.t) conn : (unit, error) result =
  let stamp = Clock.now () in
  exec conn stamp_project_deleted (stamp, stamp, project.id)

let delete_issue (issue : Issue.t) conn : (unit, error) result =
  let stamp = Clock.now () in
  exec conn stamp_issue_deleted (stamp, stamp, issue.id)

let restore_project (deleted : Project.t Deleted.t) conn : (unit, error) result =
  exec conn clear_project_deleted (Clock.now (), deleted.inner.id)

let restore_issue (deleted : Issue.t Deleted.t) conn : (unit, error) result =
  exec conn clear_issue_deleted (Clock.now (), deleted.inner.id)

(** Creation: insert, take the id back from [RETURNING], load the row. The load is what makes the
    returned object the stored one rather than a hopeful copy of the draft — a project carries
    counts and an issue carries its project's slug, and neither is on the row that was written. *)
let created what load conn id =
  let* found = load conn id in
  match found with Some row -> Ok row | None -> Error (Missing_after_insert what)

let create_project (draft : Project.draft) conn : (Project.t, error) result =
  let stamp = Clock.now () in
  let* id =
    find conn insert_project
      (draft.slug, draft.title, draft.body, Project.Status.Active, stamp, stamp)
  in
  created "the project" (fun conn id -> project_by_id id conn) conn id

let create_issue (draft : Issue.draft) conn : (Issue.t, error) result =
  let stamp = Clock.now () in
  let* id =
    find conn insert_issue
      (draft.project_id, draft.title, draft.body, Issue.Status.Todo, draft.priority, stamp, stamp)
  in
  created "the issue" issue_row conn id

File: spikes/rust/tests/store.rs
//! The SQL edge, against a real in-memory SQLite database.
//!
//! No mocks, no `MockDatabase`, and no seam invented for one: `initialise` runs
//! the same DDL the binary runs and every query below goes to sqlite. sqlx
//! opens `sqlite::memory:` shared-cache and sea-orm caps a SQLite pool at one
//! connection, so one `connect` is one database and one fixture — the tests
//! cannot see each other's rows.
//!
//! The soft-delete cases are the point. Liveness is derived rather than stored,
//! so what has to be shown is that one row written on the way out is one row
//! cleared on the way back, and that nothing else changed in between.

use tt_spike::domains::issue::services as issue_services;
use tt_spike::domains::issue::{self, Priority, Status};
use tt_spike::domains::project::services as project_services;
use tt_spike::domains::project::{self, Project};
use tt_spike::domains::schema;
use tt_spike::platform::db::{self, Db};

async fn fixture() -> Db {
    let db = db::connect("sqlite::memory:")
        .await
        .expect("connect should succeed");
    schema::initialise(&db).await.expect("DDL should run");
    db
}

async fn a_project(db: &Db, slug: &str) -> Project {
    project_services::create_project(
        db,
        project::Draft {
            slug: slug.into(),
            title: slug.into(),
            body: String::new(),
        },
    )
    .await
    .expect("create_project should succeed")
}

async fn an_issue(db: &Db, project: &Project, title: &str, priority: Priority) -> issue::Issue {
    issue_services::create_issue(
        db,
        issue::Draft {
            project_id: project.id,
            title: title.into(),
            body: String::new(),
            priority,
        },
    )
    .await
    .expect("create_issue should succeed")
}

fn titles(issues: &[issue::Issue]) -> Vec<&str> {
    issues.iter().map(|i| i.title.as_str()).collect()
}

// --- creation -------------------------------------------------------------

#[tokio::test]
async fn create_returns_the_stored_row_with_its_assigned_id() {
    let db = fixture().await;
    let tt = a_project(&db, "tt").await;
    assert!(tt.id > 0);
    assert_eq!(tt.status, project::Status::Active);
    assert_eq!((tt.todo, tt.doing, tt.done), (0, 0, 0));
    // One instant in both stamps, because a write takes the clock once.
    assert_eq!(tt.created_at, tt.updated_at);

    let other = a_project(&db, "other").await;
    assert_ne!(other.id, tt.id);

    let issue = an_issue(&db, &tt, "first", Priority::High).await;
    assert!(issue.id > 0);
    assert_eq!(issue.status, Status::Todo);
    // The projection, not the draft: `project_slug` is not a column, it comes
    // from the join.
    assert_eq!(issue.project_slug, "tt");
    assert_eq!(issue.created_at, issue.updated_at);
}

#[tokio::test]
async fn counts_are_part_of_the_projection() {
    let db = fixture().await;
    let tt = a_project(&db, "tt").await;
    for title in ["a", "b", "c"] {
        an_issue(&db, &tt, title, Priority::Normal).await;
    }
    let issues = issue_services::issues(&db, "tt")
        .await
        .expect("issues should load");
    let doing = issue::Issue {
        status: Status::Doing,
        ..issues[0].clone()
    };
    issue_services::update_issue(&db, &doing)
        .await
        .expect("update");

    let tt = project_services::project(&db, "tt")
        .await
        .expect("project should load")
        .expect("project should be there");
    assert_eq!((tt.todo, tt.doing, tt.done), (2, 1, 0));
    assert_eq!(tt.issue_count(), 3);

    // A project with nothing under it still gets zeroes rather than nothing.
    let empty = a_project(&db, "empty").await;
    let empty = project_services::project(&db, &empty.slug)
        .await
        .expect("project should load")
        .expect("project should be there");
    assert_eq!((empty.todo, empty.doing, empty.done), (0, 0, 0));
}

// --- ordering -------------------------------------------------------------

#[tokio::test]
async fn issues_come_back_high_priority_first_then_oldest_first() {
    let db = fixture().await;
    let tt = a_project(&db, "tt").await;
    // Inserted normal, high, high — so neither insertion order nor priority
    // alone produces the expected order.
    for (title, priority) in [
        ("first", Priority::Normal),
        ("second", Priority::High),
        ("third", Priority::High),
    ] {
        an_issue(&db, &tt, title, priority).await;
    }
    let issues = issue_services::issues(&db, "tt")
        .await
        .expect("issues should load");
    assert_eq!(titles(&issues), vec!["second", "third", "first"]);
}

// --- updates --------------------------------------------------------------

#[tokio::test]
async fn an_update_writes_the_editable_columns_and_leaves_the_rest() {
    let db = fixture().await;
    let tt = a_project(&db, "tt").await;
    let issue = an_issue(&db, &tt, "first", Priority::Normal).await;

    let edited = issue::Issue {
        title: "renamed".into(),
        body: "why".into(),
        status: Status::Doing,
        priority: Priority::High,
        status_note: Some("started".into()),
        ..issue.clone()
    };
    issue_services::update_issue(&db, &edited)
        .await
        .expect("update");

    let read = issue_services::issue(&db, issue.id)
        .await
        .expect("issue should load")
        .expect("issue should be there");
    assert_eq!(read.title, "renamed");
    assert_eq!(read.status_note, Some("started".into()));
    assert_eq!(read.created_at, issue.created_at);
    assert_eq!(read.project_id, issue.project_id);

    // A nullable column has one write path: `Set(None)` clears it.
    let cleared = issue::Issue {
        status_note: None,
        ..read
    };
    issue_services::update_issue(&db, &cleared)
        .await
        .expect("update");
    assert_eq!(
        issue_services::issue(&db, issue.id)
            .await
            .expect("issue should load")
            .expect("issue should be there")
            .status_note,
        None
    );
}

// --- soft deletes ---------------------------------------------------------

#[tokio::test]
async fn deleting_a_project_hides_its_issues_and_restoring_brings_them_back() {
    let db = fixture().await;
    let tt = a_project(&db, "tt").await;
    an_issue(&db, &tt, "kept", Priority::Normal).await;
    let doomed = an_issue(&db, &tt, "doomed", Priority::Normal).await;

    // One issue deleted in its own right, before the project goes.
    issue_services::delete_issue(&db, &doomed)
        .await
        .expect("delete");
    assert_eq!(
        titles(&issue_services::issues(&db, "tt").await.expect("issues")),
        vec!["kept"]
    );

    project_services::delete_project(&db, &tt)
        .await
        .expect("delete");
    assert!(
        project_services::project(&db, "tt")
            .await
            .expect("project")
            .is_none()
    );
    assert!(
        issue_services::issues(&db, "tt")
            .await
            .expect("issues")
            .is_empty()
    );
    assert!(
        issue_services::issue(&db, doomed.id)
            .await
            .expect("issue")
            .is_none()
    );

    // The trash is by the row's own `deleted_at`, so the hidden issue is not in
    // it — one row was written on the way out, not two.
    let trashed_projects = project_services::trashed_projects(&db)
        .await
        .expect("trash");
    assert_eq!(trashed_projects.len(), 1);
    assert_eq!(trashed_projects[0].inner.slug, "tt");
    assert_eq!(
        issue_services::trashed_issues(&db)
            .await
            .expect("trash")
            .iter()
            .map(|d| d.inner.title.as_str())
            .collect::<Vec<_>>(),
        vec!["doomed"]
    );

    project_services::restore_project(&db, &trashed_projects[0])
        .await
        .expect("restore");

    // Exactly the issues that were not deleted in their own right.
    assert_eq!(
        titles(&issue_services::issues(&db, "tt").await.expect("issues")),
        vec!["kept"]
    );
    assert_eq!(
        issue_services::trashed_issues(&db)
            .await
            .expect("trash")
            .len(),
        1
    );
}

#[tokio::test]
async fn an_issue_can_be_restored_on_its_own() {
    let db = fixture().await;
    let tt = a_project(&db, "tt").await;
    let issue = an_issue(&db, &tt, "back", Priority::Normal).await;
    issue_services::delete_issue(&db, &issue)
        .await
        .expect("delete");

    let trashed = issue_services::trashed_issues(&db).await.expect("trash");
    assert_eq!(trashed.len(), 1);
    // The trash row still knows its project, because that read asks nothing of
    // the project but its slug.
    assert_eq!(trashed[0].inner.project_slug, "tt");

    issue_services::restore_issue(&db, &trashed[0])
        .await
        .expect("restore");
    assert_eq!(
        titles(&issue_services::issues(&db, "tt").await.expect("issues")),
        vec!["back"]
    );
    assert!(
        issue_services::trashed_issues(&db)
            .await
            .expect("trash")
            .is_empty()
    );
}

#[tokio::test]
async fn a_slug_is_reusable_once_its_project_is_deleted() {
    let db = fixture().await;
    let first = a_project(&db, "tt").await;
    project_services::delete_project(&db, &first)
        .await
        .expect("delete");

    let second = a_project(&db, "tt").await;
    assert_ne!(second.id, first.id);
    assert_eq!(
        project_services::projects(&db)
            .await
            .expect("projects")
            .iter()
            .map(|p| p.slug.as_str())
            .collect::<Vec<_>>(),
        vec!["tt"]
    );
}

// --- the two assertions the OCaml side cannot make ------------------------

#[tokio::test]
async fn the_emitted_schema_is_the_designed_one() {
    let db = fixture().await;
    let ddl = db::emitted_ddl(&db).await.expect("sqlite_master");
    let all = ddl.join("\n");

    // `sqlite_master` keeps the statement as written apart from `IF NOT
    // EXISTS`, which it drops.
    for table in ["CREATE TABLE projects", "CREATE TABLE issues"] {
        assert!(all.contains(table), "missing {table} in:\n{all}");
    }
    assert_eq!(
        all.matches("STRICT").count(),
        2,
        "both tables STRICT:\n{all}"
    );
    assert!(all.contains("CHECK (status IN ('active', 'archived'))"));
    assert!(all.contains("CHECK (status IN ('todo', 'doing', 'done'))"));
    assert!(all.contains("CHECK (priority IN (0, 1))"));
    assert!(all.contains("projects_slug_live"));
    assert!(all.contains("issues_by_project"));
    assert!(all.contains("ON DELETE CASCADE"));

    // The entity structs declare the columns and the DDL declares the table, so
    // this is the assertion that the two describe one thing. `sea-orm`'s own
    // `Schema::create_table_from_entity` could emit none of the above, which is
    // why the DDL is written out rather than generated.
    for (table, columns) in schema::declared_columns() {
        let statement = ddl
            .iter()
            .find(|sql| sql.contains(&format!("CREATE TABLE {table}")))
            .unwrap_or_else(|| panic!("no CREATE TABLE for {table}"));
        for column in columns {
            assert!(
                statement.contains(&format!("\n        {column} ")),
                "{table}.{column} is declared by the entity and not by the DDL:\n{statement}"
            );
        }
    }
}

#[tokio::test]
async fn a_duplicate_live_slug_is_refused_by_the_constraint() {
    // `createProject`'s hook is what turns this into a sentence; the partial
    // unique index is what guarantees it. This bypasses the hook entirely, so
    // what answers is the database.
    let db = fixture().await;
    a_project(&db, "tt").await;
    let again = project_services::create_project(
        &db,
        project::Draft {
            slug: "tt".into(),
            title: String::new(),
            body: String::new(),
        },
    )
    .await;
    assert!(again.is_err(), "the index should have refused it");
}

#[tokio::test]
async fn the_enum_round_trips_through_the_column_it_is_stored_in() {
    // `DeriveActiveEnum`'s `string_value` is a second table of strings beside
    // serde's `rename_all`, and the type system relates them not at all. This
    // is the only thing that says they agree: the values go through a column
    // whose CHECK constraint spells the wire names out, and come back as the
    // same variants.
    let db = fixture().await;
    let tt = a_project(&db, "tt").await;
    for (title, status, priority) in [
        ("a", Status::Todo, Priority::Normal),
        ("b", Status::Doing, Priority::High),
        ("c", Status::Done, Priority::Normal),
    ] {
        let issue = an_issue(&db, &tt, title, priority).await;
        let moved = issue::Issue { status, ..issue };
        issue_services::update_issue(&db, &moved)
            .await
            .expect("update");
        let read = issue_services::issue(&db, moved.id)
            .await
            .expect("issue should load")
            .expect("issue should be there");
        assert_eq!(read.status, status);
        assert_eq!(read.priority, priority);
    }

    let archived = Project {
        status: project::Status::Archived,
        ..project_services::project(&db, "tt")
            .await
            .expect("project")
            .expect("project should be there")
    };
    project_services::update_project(&db, &archived)
        .await
        .expect("update");
    assert_eq!(
        project_services::project(&db, "tt")
            .await
            .expect("project")
            .expect("project should be there")
            .status,
        project::Status::Archived
    );
}
